flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java

```diff
@@ -165,6 +166,8 @@ public abstract class SchedulerBase implements SchedulerNG, CheckpointScheduling
 
     private final DeploymentStateTimeMetrics deploymentStateTimeMetrics;
 
+    private final Map<JobVertexID, BitSet> finishingTasks;
```

Comment:
Currently the per-execution-vertex information is stored in ExecutionVertex. Would it be better to also place this information in `ExecutionVertex`?

Reply:
The `tasks-end-of-data` information is only for final checkpointing, if the flink job is not streaming or without checkpoint-enabled, the `tasks-end-of-data` field can always be empty. Currently, would it be better to keep it in scheduler just like the existing `executionVertexVersioner` which is also per-execution-vertex information.

Reply:
Thanks for the update. The latest implementation LGTM.